### ISSUE TITLE
What's with double IP addresses failing INSERT into audit_log table?

### ISSUE THREAD

**Issue by pitus** [mautic/mautic#314]
This is getting frustrating, mautic logs are soooo useless, someone hasn't been thinking at all when adding error information into them with the source code/line number of the line that writes it into the sql rather than the source/line of the actual exception that's being logged ...

For example, already reported this useless error which used to happen thousands of times, not it only happens tens of times a day... But still remains totally useless - no info at all is carried by it.

`[2015-04-16 14:04:01] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Sorry! The requested URL was not found." at /www_dev/httpdocs/PROD/mautic/vendor/symfony/framework-bundle/Symfony/Bundle/FrameworkBundle/Controller/Controller.php line 151 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): Sorry! The requested URL was not found. at /www_dev/httpdocs/PROD/mautic/vendor/symfony/framework-bundle/Symfony/Bundle/FrameworkBundle/Controller/Controller.php:151)"} []`

Then, these got replaced (well, statistically at least as each was there before, just as often) with:`

`[2015-04-16 14:35:30] mautic.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\DBALException: "An exception occurred while executing 'INSERT INTO audit_log (user_id, user_name, bundle, object, object_id, action, details, date_added, ip_address) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [0, "", "lead", "lead", 11109, "create", "a:1:{s:11:\"ipAddresses\";a:2:{i:0;s:0:\"\";i:1;s:27:\"89.76.54.134, 66.249.81.220\";}}", "2015-04-16 14:35:30", "89.76.54.134, 66.249.81.220"]:  SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'ip_address' at row 1" at /www_dev/httpdocs/PROD/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php line 91 {"exception":"[object] (Doctrine\\DBAL\\DBALException(code: 0): An exception occurred while executing 'INSERT INTO audit_log (user_id, user_name, bundle, object, object_id, action, details, date_added, ip_address) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [0, \"\", \"lead\", \"lead\", 11109, \"create\", \"a:1:{s:11:\\\"ipAddresses\\\";a:2:{i:0;s:0:\\\"\\\";i:1;s:27:\\\"89.76.54.134, 66.249.81.220\\\";}}\", \"2015-04-16 14:35:30\", \"89.76.54.134, 66.249.81.220\"]:\n\nSQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'ip_address' at row 1 at /www_dev/httpdocs/PROD/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:91, PDOException(code: 22001): SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'ip_address' at row 1 at /www_dev/httpdocs/PROD/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Statement.php:165)"} []`

Where do these double-IP addresses come from? what's `89.76.54.134, 66.249.81.220` ? And why is mautic trying to write that into `audit_log.ip_address` which only accepts 15 characters? We have hundreds of such logs now (using 1.0.3 but this appears to have started just before that).

**Comment by alanhartless** [comment]
Hi,

With the 404s, Symfony is recording the message displayed to the user which is just the "Sorry! The requested URL was not found."  We'll discuss adding the URL to the message displayed to the user so that it's also recorded to the log but your apache access log should have the exact URL for the 404.

As for the audit log errors, the IP address fed to the query comes from the request and headers:

```
        $ipHolders = array(
            'HTTP_CLIENT_IP',
            'HTTP_X_FORWARDED_FOR',
            'HTTP_X_FORWARDED',
            'HTTP_X_CLUSTER_CLIENT_IP',
            'HTTP_FORWARDED_FOR',
            'HTTP_FORWARDED',
            'REMOTE_ADDR'
        );
```

It seems that one of those is returning the double IP for you.  Are you using some sort of CDN or something like Cloudeflare? If so, which one?

Thanks,
Alan

**Comment by pitus** [comment]
I'm not using any of them ... and my IP is a singular, but there are plenty of these double IPs from some other users being tracked by Mautic right now on a busy site.

**Comment by alanhartless** [comment]
Here's the code that is used to get the IP address:

```
    public function getIpAddressFromRequest()
    {
        $request = $this->getRequest();
        $ipHolders = array(
            'HTTP_CLIENT_IP',
            'HTTP_X_FORWARDED_FOR',
            'HTTP_X_FORWARDED',
            'HTTP_X_CLUSTER_CLIENT_IP',
            'HTTP_FORWARDED_FOR',
            'HTTP_FORWARDED',
            'REMOTE_ADDR'
        );

        foreach ($ipHolders as $key) {
            if ($request->server->get($key)) {
                return $request->server->get($key);
            }
        }

        // if everything else failes
        return '127.0.0.1';
    }
``` 

Your web server must be setting the double IP in one of those $_SERVER vars.  Can you do a var_dump($_SERVER) somewhere on the server to see which variable is being used?

**Comment by pitus** [comment]
Will check that shortly - have to wait for a sys-admin to run that for me, I only have sftp access to it.

**Comment by alanhartless** [comment]
If you have ftp access, you could try uploading a temp test.php file to the server with

```
<?php
echo '<pre>'.print_r($_SERVER,true).'</pre>';
```

Then hit up test.php in the browser

**Comment by pitus** [comment]
Well, I'm not sure what that will tell us as for me accessing it, I only see a single IP ... the double-IP comes from someone else hitting the site...

Here's what I got - only relevant keys...

`Array
(
    [HTTP_USER_AGENT] => Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.118 Safari/537.36
    [HTTP_X_FORWARDED_PROTO] => http
    [HTTP_X_FORWARDED_FOR] => 78.9.100.225
    [PATH] => /usr/local/bin:/usr/bin:/bin:/usr/lib/git-core
    [SERVER_SOFTWARE] => Apache
    [SERVER_ADDR] => 172.16.0.34
    [SERVER_PORT] => 80
    [REMOTE_ADDR] => 172.16.0.31
    [DOCUMENT_ROOT] => /www_dev/httpdocs/PROD/mautic
    [SCRIPT_FILENAME] => /www_dev/httpdocs/PROD/mautic/test.php
    [GATEWAY_INTERFACE] => CGI/1.1
    [SERVER_PROTOCOL] => HTTP/1.1
    [REQUEST_METHOD] => GET
    [PHP_SELF] => /test.php
)`

**Comment by alanhartless** [comment]
What's the site? I'll hit it then you can delete the file.

**Comment by alanhartless** [comment]
or you can email me at alan.hartless@mautic.org if you prefer

**Comment by pitus** [comment]
`http://m.decoraton.net/popups/test.php`

**Comment by alanhartless** [comment]
It was single for me as well. Okay so it seems that if a request is behind a proxy, HTTP_X_FORWARDED_FOR can be a comma separated list. Looking to remedy that.

**Comment by alanhartless** [comment]
@pitus Can you test this to see if it fixes the issue? https://github.com/mautic/mautic/pull/318 I don't have access to a proxy that forces multiple IPs so can't really test it although in theory should work.

**Comment by pitus** [comment]
Applied the changes to the code ... we'll see in the morning what's in the logs - I'll let you know then.

**Comment by pitus** [comment]
So far so good ... no new errors reported due to double-IPs ... only 404 Not Founds we forced manually... ;)

**Comment by alanhartless** [comment]
Closing as it should have been fixed in 1.0.4.

Thanks!
Alan
